File: Python/LeetCode/Num21.py
"""
Date:
  23.06.15
Number:
  LeetCode 21
Title:
   Merge Two Sorted Lists
Level:
  Easy
Author:
  thelight0804
"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
  def mergeTwoLists(self, list1, list2):
    """
    :type list1: Optional[ListNode]
    :type list2: Optional[ListNode]
    :rtype: Optional[ListNode]
    """
    #exception empty list
    if list1 is None: return list2
    if list2 is None: return list1
    
    #add dummyNode
    dummyNode = ListNode(-1)
    crntNode = dummyNode

    node1 = list1
    node2 = list2

    #merge two lists
    while node1 and node2:
      if node1.val <= node2.val:
        crntNode.next = node1 #crntNode link next node
        node1 = node1.next #node1 move next
        crntNode = crntNode.next #crntNode move next
      else:
        crntNode.next = node2
        node2 = node2.next
        crntNode = crntNode.next
    
    #link remain node link
    if node1:
      crntNode.next = node1
    else:
      crntNode.next = node2

    return dummyNode.next

File: JavaScript/LeetCode/Num21.js
/**
 * Date: 23.10.19
 * Number: LeetCode 21
 * Title: Merge Two Sorted Lists
 * Level: Easy
 * Author: thelight0804
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    // empty list1 or list 2
    if (!list1) return list2;
    else if (!list2) return list1;

    // create new merge list
    var list = new ListNode(-1, null);
    var crnt = list; //set current node of new merge list

    // iterate lists
    while (list1 && list2){
        // crnt.next link to list 1
        if (list1.val < list2.val){
            crnt.next = list1;
            list1 = list1.next;
        }
        // crnt.next link to list 2
        else {
            crnt.next = list2;
            list2 = list2.next;
        }
        // move crnt node to next
        crnt = crnt.next;
    }

    // link last nodes in list
    crnt.next = list1 || list2;
    
    // return merged list
    return list.next;
};